src/Core/Checkout/Promotion/Cart/PromotionCollector.php

```diff
@@ -170,7 +170,10 @@ public function collect(CartDataCollection $data, Cart $original, SalesChannelCo
             // when being in a recalculation, having notifications about the removal of automatic promotion is desired
             // addition notifications are handled as usual in the PromotionCalculator
             if ($behavior->isRecalculation()) {
-                $oldPromotions = $original->getLineItems()->filter(static fn (LineItem $item) => !$item->getReferencedId())->getElements();
+                $oldPromotions = $original->getLineItems()
+                    ->filterType(PromotionProcessor::LINE_ITEM_TYPE)
+                    ->filter(static fn (LineItem $item) => !$item->getReferencedId())
```

Comment:
I would filter this in one function. Every filter creates a new collection, seems overly resource intensive.